sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs

```diff
@@ -483,197 +494,696 @@ public virtual async Task<ServiceBusMessage> PeekAsync(CancellationToken cancell
+        }
 
-            await Task.Delay(1).ConfigureAwait(false);
+        /// <summary>Indicates that the receiver wants to defer the processing for the message.</summary>
+        ///
+        /// <param name="lockToken">The lock token of the <see cref="ServiceBusMessage" />.</param>
+        /// <param name="timeout"></param>
+        /// <param name="propertiesToModify">The properties of the message to modify while deferring the message.</param>
+        ///
+        internal virtual Task DeferAsyncInternal(
+            string lockToken,
+            TimeSpan timeout,
+            IDictionary<string, object> propertiesToModify = null)
+        {
+            var lockTokens = new[] { new Guid(lockToken) };
+            if (lockTokens.Any(lt => this.requestResponseLockedMessages.Contains(lt)))
+            {
+                return this.DisposeMessageRequestResponseAsync(lockTokens, timeout, DispositionStatus.Defered, propertiesToModify);
+            }
+            return this.DisposeMessagesAsync(lockTokens, GetDeferOutcome(propertiesToModify), timeout);
         }
 
         /// <summary>
         /// Completes a series of <see cref="ServiceBusMessage"/> using a list of lock tokens. This will delete the message from the service.
         /// </summary>
-        /// <remarks>
-        /// A lock token can be found in <see cref="ServiceBusMessage.SystemPropertiesCollection.LockToken"/>,
-        /// only when <see cref="ReceiveMode"/> is set to <see cref="ReceiveMode.PeekLock"/>.
-        /// This operation can only be performed on messages that were received by this receiver.
-        /// </remarks>
+        ///
         /// <param name="lockTokens">An <see cref="IEnumerable{T}"/> containing the lock tokens of the corresponding messages to complete.</param>
+        /// <param name="outcome"></param>
+        /// <param name="timeout"></param>
+        internal async Task DisposeMessagesAsync(
+            IEnumerable<Guid> lockTokens,
+            Outcome outcome,
+            TimeSpan timeout)
+        {
+            if (IsSessionReceiver)
+            {
+                // TODO -  ThrowIfSessionLockLost();
+            }
+
+            List<ArraySegment<byte>> deliveryTags = this.ConvertLockTokensToDeliveryTags(lockTokens);
+
+            ReceivingAmqpLink receiveLink = null;
+            try
+            {
+                ArraySegment<byte> transactionId = AmqpConstants.NullBinary;
+                //var ambientTransaction = Transaction.Current;
+                //if (ambientTransaction != null)
+                //{
+                //    transactionId = await AmqpTransactionManager.Instance.EnlistAsync(ambientTransaction, this.ServiceBusConnection).ConfigureAwait(false);
+                //}
+
+                if (!Consumer.ReceiveLink.TryGetOpenedObject(out receiveLink))
+                {
+                    receiveLink = await Consumer.ReceiveLink.GetOrCreateAsync(timeout).ConfigureAwait(false);
```

Comment:
Can the implementation be moved the AmqpConsumer? For Peek we needed to both open the receive link and use the management link. But even there, I probably should have moved the ReceiveLinkName logic into the Consumer.

Reply:
Moved implementation to the Amqp layer.